cpp/src/arrow/util/iterator_test.cc

```diff
@@ -319,8 +309,9 @@ TEST(ReadaheadIterator, Trace) {
       std::make_shared<TracingIterator<TestInt>>(VectorIt({1, 2, 3, 4, 5, 6, 7, 8}));
   ASSERT_EQ(tracing->values().size(), 0);
 
-  Iterator<TestInt> it;
-  ASSERT_OK(MakeReadaheadIterator(MakePointerIterator(tracing), 2, &it));
+  auto tracing_copy = MakeFunctionIterator(
+      internal::MemberFunction(tracing, &TracingIterator<TestInt>::Next));
```

Comment:
This is cryptic. Why is `MemberFunction` needed while `TracingIterator` is an iterator? There should be a more idiomatic way to do this.

Reply:
The problem arises from TracingIterator, which is referenced as it is being consumed in order to make calls to its assert methods. Since Iterators are move only, this means we need to wrap the iterator in a shared pointer and create another iterator which wraps that shared pointer. I'll refactor this to separate TracingIterator and TracingIteratorState